packages/common/src/audius-query/createRequestBatcher.ts

```diff
@@ -0,0 +1,140 @@
+import { omit, keyBy } from 'lodash'
+import objectHash from 'object-hash'
+
+import { ID } from '~/models'
+import { Nullable } from '~/utils/typeUtils'
+
+import { AudiusQueryContextType } from './AudiusQueryContext'
+import { schemas } from './schema'
+import { EndpointConfig } from './types'
+
+// Requests occuring in this period will be batched
+const BATCH_PERIOD_MS = 10
+
+type RequestQueueItem<Args, Data> = {
+  fetchArgs: Args
+  context: AudiusQueryContextType
+  resolve: (value: Data) => void
+  reject: (error: unknown) => void
+}
+
+type RequestGroup = {
+  queue: RequestQueueItem<any, any>[]
+  timer: Nullable<NodeJS.Timeout>
+}
+
+/**
+ * Create a request batcher that batches requests for the same endpoint
+ */
+export const createRequestBatcher = () => {
+  const requestGroups: Record<string, RequestGroup> = {}
+
+  const fetch =
+    <Args, Data>(
+      endpointName: string,
+      endpointConfig: EndpointConfig<Args, Data>
+    ) =>
+    (fetchArgs: Args, context: AudiusQueryContextType) => {
+      return new Promise<Data>((resolve, reject) => {
+        const idArgKey = endpointConfig.options.idArgKey ?? 'id'
+
+        // Hash the request args and endpointName to determine the request group
+        const requestGroupKey = objectHash({
+          args: omit(fetchArgs ?? {}, idArgKey),
+          endpointName
+        })
+
+        requestGroups[requestGroupKey] = requestGroups[requestGroupKey] ?? {
+          queue: [],
+          timer: null
+        }
+
+        const requestGroup = requestGroups[requestGroupKey]
+
+        // Queue the request
+        requestGroup.queue = [
+          ...requestGroup.queue,
+          { fetchArgs, context, resolve, reject }
+        ]
+
+        // Set up a timer to perform the batch request
+        if (!requestGroup.timer) {
+          requestGroup.timer = setTimeout(
+            () => performBatch(requestGroup, endpointName, endpointConfig),
+            BATCH_PERIOD_MS
+          )
+        }
+      })
+    }
+
+  const performBatch = async <Args, Data>(
+    requestGroup: RequestGroup,
+    endpointName: string,
+    endpointConfig: EndpointConfig<Args, Data>
+  ) => {
+    const idArgKey = endpointConfig.options.idArgKey ?? 'id'
+    const queuedRequests = [...requestGroup.queue] as RequestQueueItem<
+      Args,
+      Data
+    >[]
+
+    requestGroup.queue = []
+    requestGroup.timer = null
+
+    // Extract the IDs from the queued requests
+    const ids = queuedRequests.map(
```

Comment:
do we want to unique these?

Reply:
Good idea, although duplicates already get filtered out on the backend. But thinking about the limit stuff Andrew mentioned, it would be better make the ids unique here. Will do!